File: py/src/ticket_tracker/notify.py
"""通知层：Discord Webhook + 电脑通知（macOS 特色 + 跨平台兜底）。"""

import json
import logging
import ssl
import subprocess
import sys
import time
import urllib.error
import urllib.request

log = logging.getLogger("ticket_tracker.notify")

IS_MAC = sys.platform == "darwin"

USER_AGENT = "ticket-tracker/1.0 (+Python urllib)"


def _ctx():
    c = ssl.create_default_context()
    c.check_hostname = False
    c.verify_mode = ssl.CERT_NONE
    return c


# ----------------- Discord -----------------

def notify_discord(webhook_url, title, message, url=None, retries=3):
    """通过 Discord Webhook 推送。返回 True/False。"""
    if not webhook_url:
        return False
    if not webhook_url.startswith("https://discord.com/api/webhooks/"):
        log.warning("Discord webhook URL 格式不符，已跳过推送")
        return False
    content = "**{}**\n{}".format(title, message)
    if url:
        content += "\n👉 {}".format(url)
    payload = json.dumps({"content": content}).encode("utf-8")
    headers = {"Content-Type": "application/json", "User-Agent": USER_AGENT}
    last = None
    for i in range(retries):
        try:
            req = urllib.request.Request(
                webhook_url, data=payload, headers=headers, method="POST")
            with urllib.request.urlopen(req, timeout=15, context=_ctx()) as resp:
                if 200 <= resp.status < 300:
                    log.info("Discord 推送成功：%s", title)
                    return True
        except (urllib.error.URLError, TimeoutError) as e:
            last = e
            if i < retries - 1:
                time.sleep(3)
    log.warning("Discord 推送失败: %s", last)
    return False


# ----------------- macOS 电脑通知（带时长上限自动停） -----------------

def notify_macos(title, message, sound=True, open_url=None, duration=60):
    """弹窗 + 周期性铃声（duration 秒内）+ 语音一次 + 自动打开购票页。
    非 macOS 平台静默返回。"""
    if not IS_MAC:
        return
    # 弹窗（系统自带几秒后自动消失）
    try:
        script = ('display notification "{}" with title "{}" sound name "Glass"'
                  .format(message.replace('"', "'"), title.replace('"', "'")))
        subprocess.run(["osascript", "-e", script], check=False, timeout=10)
    except (OSError, subprocess.SubprocessError):
        pass
    # 自动打开购票页（只开一次）
    if open_url:
        try:
            subprocess.Popen(["open", open_url])
        except (OSError, subprocess.SubprocessError):
            pass
    # 周期性响铃 + 一次性语音
    deadline = time.time() + max(0, duration)
    said = False
    while sound and time.time() < deadline:
        sys.stdout.write("\a")
        sys.stdout.flush()
        try:
            subprocess.run(["afplay", "/System/Library/Sounds/Glass.aiff"],
                           check=False, timeout=10)
        except (OSError, subprocess.SubprocessError):
            pass
        if not said:
            try:
                subprocess.run(["say", "蜘蛛侠预售已开启，快去抢票"],
                               check=False, timeout=10)
            except (OSError, subprocess.SubprocessError):
                pass
            said = True
        time.sleep(3)


# ----------------- 防休眠 -----------------

# 当前 caffeinate 子进程（模块级状态，供 TUI 状态栏展示）
_caffeinate_proc = None


def caffeinate_start(child_pid):
    """开启 caffeinate 绑定 child_pid。返回 subprocess.Popen（None 表示不可用）。"""
    global _caffeinate_proc
    if not IS_MAC:
        return None
    try:
        p = subprocess.Popen(
            ["caffeinate", "-i", "-s", "-w", str(child_pid)],
            stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL)
        _caffeinate_proc = p
        return p
    except OSError as e:
        log.warning("caffeinate 启动失败（已忽略）：%s", e)
        return None


def caffeinate_stop(proc):
    global _caffeinate_proc
    if proc and proc.poll() is None:
        try:
            proc.terminate()
            proc.wait(timeout=5)
        except (OSError, subprocess.TimeoutExpired):
            try:
                proc.kill()
            except OSError:
                pass
    # 清掉引用，无论是否同一个对象，让 is_caffeinated() 反映真实情况
    if proc is _caffeinate_proc or _caffeinate_proc is None:
        _caffeinate_proc = None
    elif _caffeinate_proc is not None and _caffeinate_proc.poll() is not None:
        _caffeinate_proc = None


def is_caffeinated():
    """caffeinate 子进程是否还活着：
    - macOS：返回 True / False
    - 其他平台：返回 None（不可用，UI 显示 "n/a"）
    """
    if not IS_MAC:
        return None
    p = _caffeinate_proc
    return p is not None and p.poll() is None

File: rs/src/notify.rs
//! 通知层：Discord Webhook + macOS 电脑通知（带 caffeinate 防休眠）。
//!
//! 与 py/.../notify.py 1:1 对齐。参考：RUST_PORT.md §5.5、§5.6。

use std::process::{Child, Command, Stdio};
use std::sync::{Mutex, OnceLock};
use std::time::Duration;

use anyhow::Result;

const USER_AGENT: &str = "ticket-tracker/1.0 (+Rust reqwest)";
const DISCORD_HOST: &str = "https://discord.com/api/webhooks/";

#[cfg(target_os = "macos")]
pub const IS_MAC: bool = true;
#[cfg(not(target_os = "macos"))]
pub const IS_MAC: bool = false;

// ----------------- Discord -----------------

pub async fn notify_discord_async(
    webhook: Option<&str>,
    title: &str,
    message: &str,
    url: Option<&str>,
) -> Result<bool> {
    let Some(wh) = webhook else {
        return Ok(false);
    };
    if !wh.starts_with(DISCORD_HOST) {
        return Ok(false);
    }
    let mut content = format!("**{}**\n{}", title, message);
    if let Some(u) = url {
        content.push_str(&format!("\n👉 {}", u));
    }
    let payload = serde_json::json!({ "content": content });

    let cli = reqwest::Client::builder()
        .user_agent(USER_AGENT)
        .timeout(Duration::from_secs(15))
        .danger_accept_invalid_certs(true)
        .build()?;

    let mut last: Option<String> = None;
    for i in 0..3 {
        match cli.post(wh).json(&payload).send().await {
            Ok(r) if r.status().is_success() => return Ok(true),
            Ok(r) => last = Some(format!("HTTP {}", r.status())),
            Err(e) => last = Some(format!("req: {}", e)),
        }
        if i + 1 < 3 {
            tokio::time::sleep(Duration::from_secs(3)).await;
        }
    }
    let _ = last; // 同 Python：warn 即可
    Ok(false)
}

pub fn notify_discord(
    webhook: Option<&str>,
    title: &str,
    message: &str,
    url: Option<&str>,
) -> Result<bool> {
    let rt = tokio::runtime::Runtime::new()?;
    rt.block_on(notify_discord_async(webhook, title, message, url))
}

// ----------------- macOS 电脑通知 -----------------

/// 在 macOS 上发系统通知 + 周期性响铃 + 语音 + 自动打开购票页。
/// 非 macOS 平台静默返回。
pub fn notify_macos(title: &str, message: &str, sound: bool, open_url: Option<&str>, duration_secs: u64) {
    if !IS_MAC {
        return;
    }
    // 弹窗
    let script = format!(
        "display notification \"{}\" with title \"{}\" sound name \"Glass\"",
        message.replace('"', "'"),
        title.replace('"', "'")
    );
    let _ = Command::new("osascript")
        .arg("-e")
        .arg(&script)
        .stdout(Stdio::null())
        .stderr(Stdio::null())
        .status();
    // 自动打开购票页（只开一次）
    if let Some(u) = open_url {
        let _ = Command::new("open").arg(u).spawn();
    }
    // 周期性响铃 + 一次性语音
    if !sound {
        return;
    }
    let start = std::time::Instant::now();
    let mut said = false;
    while start.elapsed() < Duration::from_secs(duration_secs.max(1)) {
        print!("\x07");
        use std::io::Write;
        let _ = std::io::stdout().flush();
        let _ = Command::new("afplay")
            .arg("/System/Library/Sounds/Glass.aiff")
            .stdout(Stdio::null())
            .stderr(Stdio::null())
            .status();
        if !said {
            let _ = Command::new("say")
                .arg("预售已开启，快去抢票")
                .stdout(Stdio::null())
                .stderr(Stdio::null())
                .status();
            said = true;
        }
        std::thread::sleep(Duration::from_secs(3));
    }
}

// ----------------- caffeinate 防休眠 -----------------

static CAFFEINATE: OnceLock<Mutex<Option<Child>>> = OnceLock::new();

fn cell() -> &'static Mutex<Option<Child>> {
    CAFFEINATE.get_or_init(|| Mutex::new(None))
}

pub fn caffeinate_start(child_pid: u32) -> Option<u32> {
    if !IS_MAC {
        return None;
    }
    let mut guard = cell().lock().ok()?;
    // 若已有，先 terminate
    if let Some(mut c) = guard.take() {
        let _ = c.kill();
    }
    match Command::new("caffeinate")
        .args(["-i", "-s", "-w", &child_pid.to_string()])
        .stdout(Stdio::null())
        .stderr(Stdio::null())
        .spawn()
    {
        Ok(c) => {
            let pid = c.id();
            *guard = Some(c);
            Some(pid)
        }
        Err(_) => None,
    }
}

pub fn caffeinate_stop() {
    if !IS_MAC {
        return;
    }
    if let Ok(mut g) = cell().lock() {
        if let Some(mut c) = g.take() {
            let _ = c.kill();
            let _ = c.wait();
        }
    }
}

pub fn is_caffeinated() -> Option<bool> {
    if !IS_MAC {
        return None;
    }
    let mut g = cell().lock().ok()?;
    match g.as_mut() {
        Some(c) => {
            // try_wait 不卡死
            match c.try_wait() {
                Ok(None) => Some(true),
                _ => Some(false),
            }
        }
        None => Some(false),
    }
}
